server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -66,83 +90,283 @@
-        for (IndexMetaData index : indicesToClose) {
-            rtBuilder.remove(index.getIndex().getName());
+    private void sendShardCloseRequest(final ShardId shardId, @Nullable final Long taskId, @Nullable final TimeValue timeout,
+                                       final ActionListener<ReplicationResponse> listener) {
+        final ShardCloseRequest shardRequest = new ShardCloseRequest(shardId);
+        if (taskId != null) {
+            shardRequest.setParentTask(clusterService.localNode().getId(), taskId);
         }
+        if (timeout != null) {
+            shardRequest.timeout(timeout);
+        }
+        transportShardCloseAction.execute(shardRequest, listener);
+    }
+
+    /**
+     * Step 3 - Move index states from OPEN to CLOSE in cluster state for indices that are ready for closing.
+     */
+    private void finalizeClosing(final Map<Index, CloseIndexResponse.IndexResult> results,
+                                 final ActionListener<CloseIndexResponse> listener) {
+        clusterService.submitStateUpdateTask("close-indices", new ClusterStateUpdateTask(Priority.URGENT) {
+                @Override
+                public ClusterState execute(final ClusterState currentState) throws Exception {
+                    final MetaData.Builder metadata = MetaData.builder(currentState.metaData());
+                    final ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
+                    final RoutingTable.Builder routingTable = RoutingTable.builder(currentState.routingTable());
+
+                    final Set<String> closedIndices = new HashSet<>();
+                    for (Map.Entry<Index, CloseIndexResponse.IndexResult> result : results.entrySet()) {
+                        final Index index = result.getKey();
+                        try {
+                            final IndexMetaData indexMetaData = metadata.getSafe(index);
+                            if (indexMetaData.getState() != IndexMetaData.State.CLOSE) {
+                                if (result.getValue().hasFailures()) {
+                                    logger.debug("closing index {} failed, removing index block because: {}", index, result.getValue());
+                                    blocks.removeIndexBlock(index.getName(), INDEX_CLOSED_BLOCK);
+                                } else {
+                                    logger.debug("closing index {} succeed, removing index routing table", index);
+                                    metadata.put(IndexMetaData.builder(indexMetaData).state(IndexMetaData.State.CLOSE));
+                                    routingTable.remove(index.getName());
+                                    closedIndices.add(index.getName());
+                                }
+
+                                // TODO at this stage we could add a private index settings that contains the version in which the
```

Comment:
can you prototype this as well? I would be interested to see how that looks like.

Reply:
I can certainly prototype this. The goal is mainly to give a better error message when attempting to freeze a closed index that were not closed using this API. I also think it will help us when moving to replicated closed indices but I need to think a bit more about it.